cmake/modules/AddLLVM.cmake

```diff
@@ -943,26 +943,39 @@ endfunction()
 
 function(add_lit_testsuites project directory)
   if (NOT CMAKE_CONFIGURATION_TYPES)
-    cmake_parse_arguments(ARG "" "" "PARAMS;DEPENDS;ARGS" ${ARGN})
-    file(GLOB_RECURSE litCfg ${directory}/lit*.cfg)
-    set(lit_suites)
-    foreach(f ${litCfg})
-      get_filename_component(dir ${f} DIRECTORY)
-      set(lit_suites ${lit_suites} ${dir})
-    endforeach()
-    list(REMOVE_DUPLICATES lit_suites)
-    foreach(dir ${lit_suites})
-      string(REPLACE ${directory} "" name_slash ${dir})
+    cmake_parse_arguments(ARG "EXCLUDE_FROM_CHECK_ALL" "FOLDER" "PARAMS;DEPENDS;ARGS" ${ARGN})
+
+    if (NOT ARG_FOLDER)
+      set(ARG_FOLDER "Test Subdirectories")
+    endif()
+
+    # Search recursively for test directories by assuming anything not
+    # in a directory called Inputs contains tests.
+    file(GLOB_RECURSE to_process LIST_DIRECTORIES true ${directory}/*)
+    foreach(lit_suite ${to_process})
+      if(NOT IS_DIRECTORY ${lit_suite})
+        continue()
+      endif()
+      string(FIND ${lit_suite} Inputs is_inputs)
+      string(FIND ${lit_suite} Output is_output)
```

Comment:
Will this exclude anything with "Inputs" or "Output" anywhere in the path?  If so, wouldn't that be undesirable if you had a directory somewhere like "TestOutputArgs" that gets excluded accidentally?

Reply:
In theory that would potentially exclude a directory from getting a target generated for it... that said, it doesn't exclude the tests from the larger running suites.

This code just creates build system utility targets for each subdirectory under the main test folder so that we can selectively run tests without needing to manually add test directories.

Reply:
So the worst outcome of the directory with "Output" in it would be that we can't run the tests contained therein selectively? That still seems like an unexpected behavior that could have other effects. Can't we just use a regex that specifies that the directory name needs to be exclusively "Inputs" or "Outputs"? Something like: `string(REGEX MATCH /Inputs$ input_regex)`?

Reply:
It does not control executing the tests at all. It controls generating a convenience build target for running the tests.

I can change the regex, but we have no cases where this is a problem today, and it seems a bit silly to me to change cherry-picked code that has worked fine in LLVM for years when we have no cases where this is a problem.